bip-0352.mediawiki

```diff
@@ -0,0 +1,471 @@
+*** Let ''t<sub>n</sub> = sha256(ser<sub>P</sub>(ecdh_shared_secret) || ser<sub>32</sub>(n))''
+*** Compute ''P<sub>n</sub> = B<sub>spend</sub> + t<sub>n</sub>·G''
+*** For each ''output'' in ''outputs_to_check'':
+**** If ''P<sub>n</sub>'' equals ''output'':
+***** Add ''P<sub>n</sub>'' to the wallet
+***** Remove ''output'' from ''outputs_to_check'' and rescan ''outputs_to_check'' with ''n++''
+**** Else, if the wallet has precomputed labels (including the change label, if used)<ref name="precompute_labels">''' Why precompute labels?''' Precomputing the labels is not strictly necessary: a wallet could track the max number of labels it has used (call it ''M'') and scan for labels by adding ''m·G'' to ''P<sub>0</sub>'' for each label ''m'' up to ''M'' and comparing to the transaction outputs. This is more performant than precomputing the labels and checking via subtraction in cases where the number of eligible outputs exceeds the number of labels in use. In practice this will mainly apply to users that choose never to use labels, or users that use a single label for generating silent payment change outputs. If using a large number of labels, the wallet would need to add all possible labels to each output. This ends up being ''n·M'' additions, where ''n'' is the number of outputs in the transaction and ''M'' is the number of labels in the wallet. By precomputing the labels, the wallet only needs to compute ''m·G'' once when creating the labeled address and can determine if a label was used via a lookup, rather than adding each label to each output.</ref>:
+***** Compute ''m·G = output - P<sub>n</sub>''
+***** Check if ''m·G'' exists in the list of labels used by the wallet
+***** If a match is found:
+****** Add the ''P<sub>n</sub> + m·G'' to the wallet
+****** Remove ''output'' from ''outputs_to_check'' and rescan ''outputs_to_check'' with ''n++''
+***** If the label is not found, negate ''output'' and check again
+*** If no matches are found, stop
+
+==== Backup and Recovery ====
+
+Since each silent payment output address is derived independently, regular backups are recommended. When recovering from a backup, the wallet will need to scan since the last backup to detect new payments.
+
+If using a seed/seed phrase only style backup, the user can recover the wallet's unspent outputs from the UTXO set (i.e. only scanning transactions with at least one unspent taproot output) and can recover the full wallet history by scanning the blockchain starting from the wallet birthday. If a wallet uses labels or generates its change addresses using the change label, this information SHOULD be included in the backup. If the user does not know whether labels or the change label were used, it is strongly recommended they always check for the change label when recovering from backup and precompute a large number of labels (e.g. 100k labels) to use when re-scanning. This ensures that the wallet can recover all funds from only a seed/seed phrase backup.
```

Comment:
>If a wallet uses labels or generates its change addresses using the change label, this information SHOULD be included in the backup.

I think we should just always scan for change. I don't see much of a downside and it simplifies things. What if a user switches from a non-change wallet to a change wallet. Do they now have to redo the backup?

>and precompute a large number of labels (e.g. 100k labels)

I'd also argue here for a stronger statement on a default.

Reply:
this should be addressed now with your recent edits